tests/snap_test.go

```diff
@@ -0,0 +1,201 @@
+	}
+	if err != nil {
+		teardown()
+		return
+	}
+
+	// install matter-pi-gpio-commander
+	if utils.LocalServiceSnap() {
+		err = utils.SnapInstallFromFile(nil, newPath)
+	} else {
+		err = utils.SnapInstallFromStore(nil, snapName, utils.ServiceChannel)
+	}
+	if err != nil {
+		teardown()
+		return
+	}
+
+	if err = setupGPIO(); err != nil {
+		teardown()
+		return
+	}
+
+	// connect interfaces:
+	utils.SnapConnect(nil, snapName+":avahi-control", "")
+	utils.SnapConnect(nil, snapName+":bluez", "")
+	utils.SnapConnect(nil, snapName+":network", "")
+	utils.SnapConnect(nil, snapName+":network-bind", "")
+	utils.SnapConnect(nil, snapName+":custom-gpio", snapName+":custom-gpio-dev")
+
+	return
+}
+
+func getMockGPIO() (string, error) {
+	gpioChipNumber, stderr, err := utils.Exec(nil, "ls /dev/gpiochip* | sort -n | tail -n 1 | cut -d'p' -f3")
+	if err != nil || stderr != "" {
+		log.Printf("Failed to get mock gpio chip number, Error %s: %s", stderr, err)
+		return "", err
+	}
+	return gpioChipNumber, nil
+}
+
+func setupGPIO() error {
+
+	TestedGpioChip := os.Getenv(specificGpioChip)
+	TestedGpioLine := os.Getenv(specifGpioLine)
+
+	gpiochip := TestedGpioChip
+	var err error
+	if gpiochip == "" {
+		gpiochip, err = getMockGPIO()
+		if err != nil {
+			log.Printf("Failed to get mock gpio chip number: %s", err)
+			return err
+		}
+
+		log.Printf("[TEST] No specific gpiochip defined, using mockup gpio: /dev/gpiochip%s", gpiochip)
+	}
+
+	gpioline := TestedGpioLine
+	if gpioline == "" {
+		gpioline = "4"
+		log.Printf("[TEST] No specific gpio line defined, using default: %s", gpioline)
+	}
+
+	utils.SnapSet(nil, snapName, "gpiochip", gpiochip)
+	utils.SnapSet(nil, snapName, "gpio", gpioline)
+
+	return nil
+}
+
+func TestBlinkOperation(t *testing.T) {
+	log.Println("[TEST] Standard blink operation")
+
+	// test blink operation
+	t.Run("test-blink", func(t *testing.T) {
+		ctx, cancel := context.WithDeadline(context.Background(), <-time.After(20*time.Second))
+		defer cancel()
+
+		stdout, _, err := utils.ExecContextVerbose(nil, ctx, snapName+".test-blink")
+		t.Logf("err: %s", err)
+		t.Logf("stdout: %s", stdout)
+		assert.Error(t, err, "Expected an error")
+		assert.Equal(t, "context deadline exceeded", err.Error())
+	})
+}
+
+func TestWifiCommander(t *testing.T) {
+	log.Println("[TEST] Wifi commander")
+
+	// install chip-tool
+	err := utils.SnapInstallFromStore(t, chipToolSnap, utils.ServiceChannel)
+	if err != nil {
+		t.Fatalf("Failed to install chip-tool: %s", err)
+	}
+	// chip-tool interfaces
+	utils.SnapConnect(nil, chipToolSnap+":avahi-observe", "")
+
+	utils.SnapStart(t, snapName)
+
+	// comission chip-tool
+	t.Run("Commission", func(t *testing.T) {
+		utils.Exec(t, "sudo chip-tool pairing onnetwork 110 20202021")
+	})
+
+	t.Run("Control", func(t *testing.T) {
+		for i := 0; i < 10; i++ {
+			time.Sleep(1 * time.Second)
+			utils.Exec(t, "sudo chip-tool onoff toggle 110 1")
+		}
+	})
```

Comment:
This isn't checking if the application actually received this command (i.e. by looking into the logs or checking GPIO state).

Reply:
Working on it! I tried to assert using the [golang gpiod library](https://pkg.go.dev/github.com/warthog618/gpiod), but once I use the functions to measure the gpio line value, it seems to interfere in the way that the lighting app works. Because the matter application starts to fail locally (a thing that wasn't happening before)

Reply:
Solved! Checked the application logs!